lib/vfscore/automount.c

```diff
@@ -0,0 +1,312 @@
+	unsigned long flags;
+	/* Mount options */
+	const char *opts;
+};
+
+static inline int vfscore_mount_volume(struct vfscore_volume *vv)
+{
+	return mount(vv->sdev, vv->path, vv->drv, vv->flags, vv->opts);
+}
+
+#ifdef CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS
+
+static const char *rootfs   = CONFIG_LIBVFSCORE_ROOTFS;
+
+#ifndef CONFIG_LIBVFSCORE_ROOTDEV
+static const char *rootdev  = "";
+#else
+static const char *rootdev  = CONFIG_LIBVFSCORE_ROOTDEV;
+#endif
+
+#ifndef CONFIG_LIBVFSCORE_ROOTOPTS
+static const char *rootopts = "";
+#else
+static const char *rootopts = CONFIG_LIBVFSCORE_ROOTOPTS;
+#endif
+
+#ifndef CONFIG_LIBVFSCORE_ROOTFLAGS
+static __u64 rootflags;
+#else
+static __u64 rootflags      = (__u64) CONFIG_LIBVFSCORE_ROOTFLAGS;
+#endif
+
+UK_LIB_PARAM_STR(rootfs);
+UK_LIB_PARAM_STR(rootdev);
+UK_LIB_PARAM_STR(rootopts);
+UK_LIB_PARAM(rootflags, __u64);
+
+#endif /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
+
+#ifdef CONFIG_LIBVFSCORE_FSTAB
+
+#define LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP			':'
+
+static char *vfscore_fstab[CONFIG_LIBVFSCORE_FSTAB_SIZE];
+
+UK_LIBPARAM_PARAM_ARR_ALIAS(fstab, &vfscore_fstab, charp,
+			    CONFIG_LIBVFSCORE_FSTAB_SIZE,
+			    "VFSCore Filesystem Table");
+
+/**
+ * Expected command-line argument format:
+ *	vfs.fstab=[
+ *		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
+ *		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
+ *		...
+ *	]
+ * These list elements are expected to be separated by whitespaces.
+ * Mount options and flags are optional.
+ */
+static char *vfscore_fstab_next_volume_arg(char **pos)
+{
+	char *ipos;
+
+	UK_ASSERT(pos);
+
+	if (!*pos)
+		return NULL;
+
+	ipos = *pos;
+	*pos = strchr(ipos, LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP);
+
+	if (*pos && **pos != '\0') {
+		**pos = '\0';
+		(*pos)++;
+	} else {
+		*pos = NULL;
+	}
+
+	return ipos;
+}
+
+static void vfscore_fstab_fetch_volume_args(char *v, struct vfscore_volume *vv)
+{
+	const char *flags;
+	char *pos;
+
+	UK_ASSERT(v);
+	UK_ASSERT(vv);
+
+	pos = v;
+
+	/* sdev, path and drv are mandatory */
+	vv->sdev = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->sdev);
+
+	vv->path = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->path);
+
+	vv->drv = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->drv);
+
+	if (!pos)
+		return;
+
+	/* Allow empty flags: "<src_dev>:<mntpoint>:<fsdriver>::<opts>" */
+	flags = vfscore_fstab_next_volume_arg(&pos);
+	if (flags && flags[0] != '\0')
+		vv->flags = strtol(flags, NULL, 0);
+	else
+		vv->flags = 0;
+
+	vv->opts = vfscore_fstab_next_volume_arg(&pos);
+	if (vv->opts && vv->opts[0] == '\0')
+		vv->opts = NULL;
+}
+#endif /* CONFIG_LIBVFSCORE_FSTAB */
+
+#ifdef CONFIG_LIBUKCPIO
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv)
+{
+	struct ukplat_memregion_desc *initrd;
+	int rc;
+
+	UK_ASSERT(vv);
+
+	/* TODO: Support multiple Initial RAM Disks */
+	rc = ukplat_memregion_find_initrd0(&initrd);
+	if (unlikely(rc < 0)) {
+		uk_pr_crit("Could not find an initrd!\n");
+
```

Comment:
I think these spaces between message and return code is not necessary for these small if blocks. I leave it up to you if you want to keep them or not.